src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -200,6 +309,23 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
   memset(ops, 0, sizeof(ops));
   grpc_op* op = ops;
   if (is_start) {
+    op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
+    op->data.recv_status_on_client.trailing_metadata = nullptr;
+    op->data.recv_status_on_client.status = &client->handshake_status_code;
+    op->data.recv_status_on_client.status_details =
+        &client->handshake_status_details;
+    op->flags = 0;
+    op->reserved = nullptr;
+    op++;
+    GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+    gpr_ref(&client->refs);
+    grpc_call_error call_error =
```

Comment:
For the client's first handshake message, don't we endup calling TSI next callback twice? One is in `on_status_received` and the other in `on_handshaker_service_resp_recv`?

Reply:
The client's first handshake message response, in successful case, should result in a received message having no `result` field, and an OK status. In this case, `maybe_complete_tsi_next` will not wait for the `RECV_STATUS` op to complete.

Eventually, the last handshaker message will arrive, having at least either non-OK status or a filled in `result` field. In this case, we'll wait for the handshake RPC's status to be received, and then invoke the TSI next callback just once.

Lastly, note that if the RECV_STATUS op completes while we don't have a TSI next callback outstanding (e.g. security handshaker is waiting for a read from remote peer), then we <i>won't</i> have a non-null `pending_recv_message_result`, and so we won't accidentally invoke the TSI next callback an extra time

Reply:
In the last handshaker message, why do we wait for the handshake RPC's status to be received?

Reply:
Please also leave this conversation open.

Reply:
> In the last handshaker message, why do we wait for the handshake RPC's status to be received?

My motivation for this is to mainly just to try to be sure that we continue to drive the handshake RPC forward until it fully completes. This is basically trying to leverage the fact that there is an outstanding TSI next call (which ultimately suggests that the associated subchannel hasn't yet connected, and that someone <i>should</i> be polling on the relevant file descriptors of this handshake RPC), to make sure that we finish the handshake RPC in a timely manner.

In a degenerate case, suppose that the handshake server sends a reply message, and so the RECV_MESSAGE op completes, but the data on the wire hasn't yet arrived to complete the RECV_STATUS op. If we completed the TSI handshake at this point, then the subchannel could become ready, the application could quickly complete its RPC, and then the application could stop polling on the completion queue of its RPC (if it e.g. only wanted to do a single RPC). The data necessary to complete the RECV_STATUS op on the wire could then show up just after, but because we don't have anyone guaranteed to be polling on the relevant file descriptors of the handshake RPC (except for the optional backup poller, which runs on a fixed time interval), the handshake RPC may just hang there and never "complete".

I.e., this provides a kind of guarantee that all handshake RPCs are either driven forward until full completion in a timely manner, or be cancelled (due to handshake failure or subchannel connection timeout).

Reply:
Thanks much for the explanation. It makes sense to me.

Reply:
In xds_cleint.cc (https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/xds/xds_client.cc), we also applied the similar model, i.e., scheduling a `grpc_closure` upon receiving message `RECV_MESSAGE` and status `RECV_STATUS`.